### PR TITLE
Preserved request metadata in response

### LINKED ISSUE #1719: Make it harder to use ExampleGet and connected Reponse factory methods
Currently `Response` can be easily created with `ExampleGet`Request metadata, as manifested in #1718.

It would be better if the api itself could prevent such bugs (e.g. by moving this into some testing module).

### PR DESCRIPTION
**Pull request by Pask423**
closes #1719 #2106

### CODE REVIEW

File: core/src/test/scalajvm/sttp/client4/testing/TestResponse.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package sttp.client4.testing
+
+import sttp.client4.Response
+import sttp.model._
+
+private[sttp] object TestResponse {
```
Comment (adamw) [inline core/src/test/scalajvm/sttp/client4/testing/TestResponse.scala:6]: maybe let's call it `ResponseStub`, keeping the terminology consistent with `BackendStub`? Also, this should be public?

File: core/src/main/scala/sttp/client4/Response.scala [hunk 2]
```diff
@@ -31,38 +31,28 @@ case class Response[+T](
     s"Response($body,$code,$statusText,${Headers.toStringSafe(headers)},$history,$request)"
 }
 
-object Response {
+private[sttp] object Response {
 
-  /** Mainly useful in tests using [[sttp.client4.testing.SttpBackendStub]], when creating stub responses.
-    */
-  val ExampleGet: RequestMetadata = new RequestMetadata {
+  val emptyGet: RequestMetadata = new RequestMetadata {
```
Comment (adamw) [inline core/src/main/scala/sttp/client4/Response.scala:36]: I guess we should remove all of these methods from `Response` once `TestResposne` is there?

Comment (adamw) [reply core/src/main/scala/sttp/client4/Response.scala:36]: ah sorry, we should only keep the helper methods, but with `requestMetadata: RequestMetadata`. We should move `emptyGet`, though, which kind of is the point of the issue that is being fixed

File: core/src/main/scala/sttp/client4/testing/AbstractBackendStub.scala [hunk 3]
```diff
@@ -77,7 +77,7 @@ abstract class AbstractBackendStub[F[_], P](
     def thenRespond[T](body: T, statusCode: StatusCode): Self = thenRespond(Response[T](body, statusCode))
     def thenRespond[T](resp: => Response[T]): Self = {
       val m: PartialFunction[GenericRequest[_, _], F[Response[_]]] = {
-        case r if p(r) => monad.eval(resp)
+        case r if p(r) => monad.eval(resp.copy(request = r.onlyMetadata))
```
Comment (adamw) [inline core/src/main/scala/sttp/client4/testing/AbstractBackendStub.scala:80]: would be good to have a test for this

File: core/src/test/scalajvm/sttp/client4/testing/TestResponse.scala (hunk 1 above)
Comment (Pask423) [reply core/src/test/scalajvm/sttp/client4/testing/TestResponse.scala:6]: yeah good idea with naming, as for public hmmm not sure if we want it to be expose to users ?

File: core/src/main/scala/sttp/client4/Response.scala (hunk 2 above)
Comment (Pask423) [reply core/src/main/scala/sttp/client4/Response.scala:36]: there is a problem here with the methods here as we use them in AbstractStubBackend when request.metadata is not computed yet, I think that we will have to leave it 
image

emptyGet should not be available outside sttp package as Respone is package scope, I thought it should be ok from user perspective

File: core/src/test/scalajvm/sttp/client4/testing/TestResponse.scala (hunk 1 above)
Comment (adamw) [reply core/src/test/scalajvm/sttp/client4/testing/TestResponse.scala:6]: yeah we do, it was exposed so far, and the idea is to use it when defining the behavior of a backend stub

File: core/src/main/scala/sttp/client4/Response.scala (hunk 2 above)
Comment (adamw) [reply core/src/main/scala/sttp/client4/Response.scala:36]: `emptyGet` or whatever the name should still live in `ResponseStub`/`TestResponse`. The point of the issue is that this shouldn't be exposed in `Response`, but explicitly marked as a test-only thing. In a backend stub, we can of course use a `ResponseStub`

**Comment by adamw** [comment]
Great, thanks!

**Pull request merged**